### PR TITLE
Suitably dynamic versioning

### PR DESCRIPTION
**Pull request by dwaynepryce**
# Suitably dynamic versioning

The following versioning code bypasses a few problems with python module versions.  The following scenarios are plausible:
- A user clones `graspologic` and runs `pip install -r requirements.txt` then executes `python` in the project directory, accessing the graspologic library by python's local folder structure.
- A users clones `graspologic` and runs `python setup.py install` in the environment of their choice, accessing the graspologic library either by the local folder structure or the .egg in their site-packages, depending on their current working directory.
- A user clones no repository and wants to install the library solely via `pip` via the `pip install ...` command, which has 2 wings to consider:
  - The user wishes to try out the latest prerelease, which is going to be published with a X.Y.ZdevYYYYMMDDBUILDNUMBER style version and can be installed via `pip install graspologic --pre`
   - The user wishes to try out the latest release, which will be published as `X.Y.Z`.

This PR supports those 4 cases (notably, it does not support `pip install .` from the root project directory, which does some super weird stuff and I gave up on trying to solve it a long time ago)

The concept is this: the actual version upon a **build action**, which can be undertaken by:
- CI building a snapshot build
- CI building a release build
- Local user building a local build

These states all require the same thing: a materialized version in a file.  This version should be created at the time of this build action.

In the case of CI, we can populate the file in our CI build process and move on.  It's the case of not being in CI where we need to consider what to do next, which leaves Local user building a local build (and local user using the filesystem as the library).

In these cases, the solution is the following: if we have a populated version.txt file, we use it. If we do not, we materialize a new version based on the `__semver` in version.py and the current time in YYYYMMDDHHmmSS format. This means that if you are running on the filesystem, and you say `import graspy; print(graspy.__version__);`, it will actually tell you the version is `0.1.0dev20200926120000` as an example.  However, when you close the interpreter and do it again, it will tell you that the version is `0.1.0dev20200926120500` - because it will create a version for you at the time of import.

However, if you were to run `python setup.py install`, the setup.py file actually takes it on itself to either get a version number or use the materialized version described above, then to write it to version.txt.  Which means that installing the graspologic library from setuptools will actually lock in the version number in perpetuity.

## Gotchas
- version.txt must always be empty in the source tree
- `pip install .` does some weird thing where it registers an entry in site-packages that is like a symlink to the local filesystem anyway so it doesn't actually make an egg which means you get a new version each time and I gave up caring at this point since we got the three primary use cases: developers, users of pre-releases, and users of releases all covered. Users who install by cloning and running pip install are just going to get a weird behavior that probably isn't that important to track down, and regardless they'll get a clear `X.Y.Zdev<timestamp>` in their `graspologic.__version__` which is enough for us to go on if there are any issues raised.

### CODE REVIEW

File: graspy/version/version.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import pkg_resources
+
+__all__: List[str] = ["version", "name"]
+
+name = "graspy"  # TODO: #454 Update in https://github.com/microsoft/graspologic/issues/454
+
+# manually updated
+__semver = "0.1.0"
+#  full version (may be same as __semver on release)
+__version_file = "version.txt"
```
Comment (j1c) [inline graspy/version/version.py:15]: Need `recursive-include gaspologic/version version.txt` included in `MANIFEST.in` since this files not included in installation. I know `package_data` in setup.py is supposed to handle it, but using `package_data` never really worked for me

File: setup.py [hunk 2]
```diff
@@ -4,16 +4,29 @@
 import os
 import sys
 from setuptools import setup, find_packages
-from sys import platform
+from typing import Tuple
+
+
+def package_metadata() -> Tuple[str, str]:
+    sys.path.insert(0, os.path.join("graspy", "version"))  # TODO: #454 Change path in https://github.com/microsoft/graspologic/issues/454
+    from version import name, version
```
Comment (loftusa) [inline setup.py:12]: out of curiosity, why the within-function import instead of importing into the global namespace?

**Comment by dwaynepryce** [comment]
This is a result of an annoying circular situation.  Setup.py is running, but not necessarily installing – which means the list of all the dependencies hasn’t actually been installed yet.

So if we were to import graspy here, it would run the `__init__.py` module loading files, which import our packages, which have their own imports, but the imports they pull haven’t been installed yet.

So we treat version as if it were a standalone module, and the only dependencies we rely on are the ones that come straight from setuptools anyway (and if you don’t have setuptools, you can’t even run setup.py in the first place, so yeah).

In other words, super minimal bootstrap library available, everything else is standard issue python, no other dependencies.

Yes, this is annoying.

Someday I’ll get some free time and get to look into pyproject.toml and see if it can smooth out some of this highly choreographed dancing we have to do ☹

**Review by j1c — APPROVED** [review]
